tests/test_examples.py

```diff
@@ -34,39 +36,72 @@ def get_file_paths():
     return test_files
 
 
-def create_checker(test_file, checker_name):
-    """Creates a test function from a test file, and a checker name.
-    test_file: The full path (string) to the file.
-    checker_name: The hyphenated checker name that should be detected.
-    An example of a valid checker_name would be: 'no-init-classes'
+def _get_full_pylint_json_output() -> List[Dict]:
+    output = subprocess.run(
+        ['pylint', '--reports=n',
+         '--rcfile=python_ta/.pylintrc',
+         '--output-format=json',
+         *get_file_paths()],
+        stderr=subprocess.STDOUT,
+        stdout=subprocess.PIPE)
+
+    jsons_output = output.stdout.decode('UTF-8')
+    jsons_output = _strip_trailing_asni(jsons_output)
+    return json.loads(jsons_output)
+
+
+def _strip_trailing_asni(str_to_strip: str) -> str:
+    """Return the string stripped to the final ']'
+
+    Running Pytest from a Python Console produces a trailing ANSI '\x1b[0m' code whereas
+    the CLI Pytest does not.
+    """
+    trailing_token = str_to_strip.split(']')[-1]
+    return str_to_strip[:-len(trailing_token)]
+
+
+def _compile_message_symbols_per_file(pylint_output_list: List[Dict]) -> Dict[str, Set[str]]:
+    """Return a dict mapping file name to the set of Pylint message symbols it raises.
+
+    The file name is the basename, not the full or relative path
+    (this is to avoid being unable to access a value because of \\ and / paths)
     """
-    # The following are captured when this function is created.
-    def new_test_func():
-        found_pylint_message = False
-        output = subprocess.run(
-            ['pylint', '--reports=n',
-             '--rcfile=python_ta/.pylintrc',
-             test_file],
-            stderr=subprocess.STDOUT,
-            stdout=subprocess.PIPE)
-        for line in output.stdout.decode('utf-8').split('\n'):
-            if checker_name in line:
-                found_pylint_message = True
-                break
-        if not found_pylint_message:
-            print('Failed: ' + test_file)  # test doesn't say which file
-        assert found_pylint_message
-    return new_test_func
+    symbols_by_file = {}
+
+    for pylint_message_data in pylint_output_list:
+        file_name = os.path.basename(pylint_message_data['path'])
+        symbol = pylint_message_data['symbol']
+
+        if file_name not in symbols_by_file:
+            symbols_by_file[file_name] = set()
+
+        symbols_by_file[file_name].add(symbol)
+
+    return symbols_by_file
+
+
+@pytest.fixture(scope='session', autouse=True)
+def symbols_by_file() -> Dict[str, Set[str]]:
+    """Run pylint on all the example files return the map of file name to the
+    set of Pylint messages it raises."""
+    pylint_list_output = _get_full_pylint_json_output()
```

Comment:
In this particular case I find the function more readable if you write all of the code in this function rather than splitting it up.

Also you can simplify the grouping by using [`itertools.groupby`](https://docs.python.org/3/library/itertools.html#itertools.groupby) (but note that you'll still need to convert the returned value from this function into a `dict`).